ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs

```diff
@@ -0,0 +1,312 @@
+      PTx.Map.insert newKey v $
+        PTx.Map.delete oldKey m
+    Nothing -> m
+
+{- | Replace all tokens with the given currency symbol from a value with a different currency symbol
+Does nothing if the CS to replace is not found
+-}
+replaceTokenByCS :: CurrencySymbol -> CurrencySymbol -> Value -> Value
+replaceTokenByCS oldCS newCS = Value . replaceIfPresent oldCS newCS . getValue
+
+-- | Remove all tokens with the given currency symbol from a value
+removeTokenByCS :: CurrencySymbol -> Value -> Value
+removeTokenByCS cs v =
+  Value $ v & getValue & PTx.Map.delete cs
+
+-- | Replace an @ScriptHash@ credentials with a new one if a match is found
+updateScriptCredential :: ScriptHash -> ScriptHash -> Credential -> Credential
+updateScriptCredential oldScriptHash newScriptHash cred
+  | cred == ScriptCredential oldScriptHash = ScriptCredential newScriptHash
+  | otherwise = cred
+
+-- | Replace an @ScriptHash@ credentials for input with a provided one (if present)
+replaceInput :: ScriptHash -> ScriptHash -> TxInInfo -> TxInInfo
+replaceInput oldScriptHash newScriptHash =
+  over
+    (O.txOut . O.address . O.credential)
+    (updateScriptCredential oldScriptHash newScriptHash)
+
+-- | Replace an @ScriptHash@ for withdrawal with a provided one (if present)
+replaceWdrl :: ScriptHash -> ScriptHash -> PTx.Map.Map StakingCredential Integer -> PTx.Map.Map StakingCredential Integer
+replaceWdrl oldScript newScript =
+  let
+    oldCredential = StakingHash $ ScriptCredential oldScript
+    newCredential = StakingHash $ ScriptCredential newScript
+   in
+    replaceIfPresent oldCredential newCredential
+
+-----------------------------------------------------------------------
+-- Attacks
+
+attackRefInputNotPresent :: Reader ScriptsTestsParams (Endo (YieldingRedeemer, ScriptContext))
+attackRefInputNotPresent =
+  pure . Endo . fmap $ set (O.txInfo . O.referenceInputs) []
+
+attackRefInputNotAuthorised :: Reader ScriptsTestsParams (Endo (YieldingRedeemer, ScriptContext))
+attackRefInputNotAuthorised = do
+  AuthorisedScriptsSTCS authorisedScriptsSTCS' <- asks authorisedScriptsSTCS
+  pure $
+    Endo $
+      fmap $
+        over
+          (O.txInfo . O.referenceInputs . traversed . O.txOut . O.value)
+          (removeTokenByCS authorisedScriptsSTCS')
+
+attackAuthorisedScriptIndexInvalid ::
+  Reader ScriptsTestsParams (Endo (YieldingRedeemer, ScriptContext))
+attackAuthorisedScriptIndexInvalid =
+  redeemerAttack
+    (O.authorisedScriptIndex . _Wrapped)
+    -- NOTE: this is a big number that will be not present in
+    -- any script context created for testing purposes
+    (const 42)
+
+attackAuthorisedProofIndexInvalidIndex ::
+  Reader ScriptsTestsParams (Endo (YieldingRedeemer, ScriptContext))
+attackAuthorisedProofIndexInvalidIndex =
+  redeemerAttack
+    (O.authorisedScriptProofIndex . _Wrapped . _2)
+    -- NOTE: this is a big number that will be not present in
+    -- any script context created for testing purposes
+    (const 42)
```

Comment:
A more stable way to do this (which would survive property testing, if requested) is to just take length (don't even need to add 1 because it's 0 indexed)